search: remove redundant len check

rvantonder opened the pull request:
Stacked on https://github.com/sourcegraph/sourcegraph/pull/23941.

rvantonder commented on internal/search/zoekt/indexed_search.go:
```diff
@@ -428,10 +428,6 @@ func doZoektSearchGlobal(ctx context.Context, q zoektquery.Q, typ search.Indexed
 
 // zoektSearch searches repositories using zoekt.
 func zoektSearch(ctx context.Context, args *search.ZoektParameters, repos *IndexedRepoRevs, since func(t time.Time) time.Duration, c streaming.Sender) error {
-	if len(repos.repoRevs) == 0 {
-		return nil
-	}
-
```
This check is redundant, we already check it before calling `zoektSearch`: 

https://github.com/sourcegraph/sourcegraph/blob/a6108388eaef9db863c607362ea120c31ea62939/internal/search/zoekt/indexed_search.go#L192-L194

which calls

https://github.com/sourcegraph/sourcegraph/blob/a6108388eaef9db863c607362ea120c31ea62939/internal/search/zoekt/indexed_search.go#L174-L179

stefanhengl reviewed (commented):
My concerns are similar to what I wrote in #23940. Would be interested what the others think about it @camdencheek @keegancsmith?

keegancsmith reviewed (commented):
This one I feels different to the nil check in the other PR. Skipping work when repoRevs is zero feels like a local optimization. Coupling it to the control flow of the caller seems brittle. I'd rather leave this in to be honest.

Unless it is a mistake to call with zero repoRevs. Then we need a way to try avoid doing that. I assume it is a mistake due to the higher level plan stuff you introduced?

rvantonder commented:
> This one I feels different to the nil check in the other PR.

Agree, it is different, also PR description was not accurate. I updated it.

> Unless it is a mistake to call with zero repoRevs. I assume it is a mistake due to the higher level plan stuff you introduced?

I wouldn't call it a mistake. I would say the interface that ensures this property/optimization isn't clearly separated. My read on this is: "We have this indexed repos request object that we construct, and checks this property for us. Separately, we want to keep a related check on whether repos are empty in `zoektSearch` because it might be called from somewhere else, and we can't guarantee at the interface level that it may be called with zero repos, and this is brittle." To which I say "Fair". But the underlying issue here to me seems that this indexed search request object really isn't that useful if we are concerned about it being bypassed to a `zoektSearch` call. There isn't a separation of concerns here. I would expect either: 

- `IndexedSearchRequest.Search` should always be a precursor to `zoektSearch` and ensures this optimization for us, or,
- `IndexedSearchRequest` shouldn't have the business of ensuring this optimization for us when it constructs its object. That business should be delegated to "thing that ensures optimization for us".  So it could go the other way, where that "thing" could be the check in `zoektSearch`, or in a function or routine that precurses it, and so maybe `IndexedSearchRequest.Search` isn't the method to ensure this optimization.

Checking in both places is a catch-all that is sometimes redundant, and reveals that we're basically OK with there being no relation between `IndexedSearchRequest.Search` and `zoektSearch`. Thsi confuses me, and if I didn't know the codebase I wouldn't know what to use to run a zoekt search. 

I honestly can't say what the _intent_ is of these two functions, and whether `IndexedSearchRequest.Search` should always precurse `zoektSearch` (this seems kind of reasonable, but more likely we are comfortable with people calling `zoektSearch` directly if they know what they're doing?). So: I've changed it to move the check to `zoektSearch` (optimization ensured at the lowest call) and update `IndexedSearchRequest` appropriately: https://github.com/sourcegraph/sourcegraph/pull/23943/commits/9173592e26c16e0cb6b1705b30565b1fdfc7384c

I'm also OK with leaving this as-is, I don't think I need it. But since we're here... let me know what you think about the change.

rvantonder commented:
I believe this change is a correct one for the concerns at hand. But, I've decided to abandon this in favor of another direction that "promotes" the use of `Repos()` so that I can express an interface across both global and non-global searches. I will revisit this if/when I have time.
